docs/user/config-file-glossary.rst
####################
Config file glossary
####################

This page is a glossary of different settings in the ``env.example`` configuration file.
All values shown are the default settings.
This glossary is intended for advanced users.


************
IRC settings
************

Host connection settings
========================

``IRC_HOST_IP=""``
    Specify IP address to use for IRC connection.
    Useful for hosts with multiple IP addresses.

Server connection settings
==============================

``IRC_SERVER=chat.freenode.net``
    IRC server to connect to

.. CAUTION:: Required setting (cannot be empty)

``IRC_SERVER_PASSWORD=""``
    IRC server password

``IRC_PORT=6697``
    IRC server port

Encryption options
------------------

``IRC_USE_SSL=false``
    Connect to the IRC server with SSL when set to `true`

``IRC_CERT_ALLOW_EXPIRED=true``
    Allow connecting to IRC server with an expired TLS/SSL certificate

``IRC_CERT_ALLOW_SELFSIGNED=true``
    Allows TeleIRC to accept TLS/SSL certificates from non-trusted/unknown Certificate Authorities (CA)

Channel settings
================

``IRC_CHANNEL="#channel"``
    IRC channel for bot to join

.. CAUTION:: Required setting (cannot be empty)

``IRC_CHANNEL_KEY=""``
    IRC channel key, for password-protected channels

``IRC_BLACKLIST=""``
    Comma-separated list of IRC nicks to ignore

Bot settings
============

``IRC_BOT_NAME=teleirc``
    IRC nickname for bot.
    Most IRC clients and bridges show this nickname.

.. CAUTION:: Required setting (cannot be empty)

``IRC_BOT_REALNAME="Powered by TeleIRC <github.com/rubensadev/teleirc>"``
    IRC ``REALNAME`` for bot.
    Often visible in IRC ``/whois`` reports, but most clients do not show this information by default.

``IRC_BOT_IDENT=teleirc``
    Identification metadata for bot connection.
    This is rarely used or shown in most IRC clients and bridges, so only change this if you know what you are doing.
    Often visible in IRC ``/whois`` reports, but most clients do not show this information by default.

NickServ options
----------------

``IRC_NICKSERV_SERVICE=NickServ``
    IRC service used for authentication.

``IRC_NICKSERV_USER=""``
    IRC NickServ username.

``IRC_NICKSERV_PASS=""``
    IRC NickServ password.

Message settings
================

``TELEGRAM_MESSAGE_PREFIX="<"``
    Text displayed before Telegram name in IRC

``TELEGRAM_MESSAGE_SUFFIX=">"``
    Text displayed after Telegram name in IRC

``IRC_SEND_STICKER_EMOJI=true``
    Send emojis associated with a sticker to IRC (when a Telegram user sends a sticker)

``IRC_SEND_DOCUMENT=true``
    Send documents and files from Telegram to IRC (`why is this true by default? <https://github.com/rubensadev/teleirc/issues/115>`_)

``IRC_EDITED_PREFIX="[EDIT] "``
    Prefix to prepend to messages when a user edits a Telegram message and it is resent to IRC

``IRC_MAX_MESSAGE_LENGTH=400``
    Maximum length of the message that can be sent to IRC.
    Longer messages are split into multiple messages.

``IRC_SHOW_ZWSP=true``
    Prevents users with the same Telegram and IRC username from pinging themselves across platforms.

``IRC_SHOW_JOIN_MESSAGE=true``
    Show in IRC when someone joins the Telegram chat.

``IRC_SHOW_LEAVE_MESSAGE=true``
    Show in IRC when someone leaves the Telegram chat.

``IRC_SHOW_LOCATION_MESSAGE=false``
    If a user shares their location on Telegram, this will forward the GPS coordinates of their location to IRC if set to true.

``IRC_NO_FORWARD_PREFIX="[off]"``
    A string users can prefix their message with to prevent it from being relayed across the bridge.
    Removing this option or setting it to "" disables it.

``IRC_QUIT_MESSAGE="TeleIRC bridge stopped."``
    A string that TeleIRC sends to the IRC channel when the application exits using IRC's "QUIT" command.
    If not specified, it simply closes the connection without providing a reason.
    The bot must be connected to a server for a certain amount of time for the server to send the quit message to the channel.


*****************
Telegram settings
*****************

``TELEGRAM_CHAT_ID=``
    Telegram chat ID of bridged group (:ref:`how do I get this? <chat-id>`). Like: `-0000000000000`

.. CAUTION:: Required setting (cannot be empty)

``TELEIRC_TOKEN=``
    Private API token for Telegram bot. Like: `000000000:AAAAAAaAAa2AaAAaoAAAA-a_aaAAaAaaaAA`

.. CAUTION:: Required setting (cannot be empty)

``MAX_MESSAGE_PER_MINUTE=20``
    Maximum number of messages sent per minute from IRC to Telegram.

``IRC_PREFIX="<"``
    Text displayed before IRC nickname in messages sent to Telegram

``IRC_SUFFIX=">"``
    Text displayed after IRC nickname in messages sent to Telegram

``TELEGRAM_MESSAGE_REPLY_PREFIX="["``
    Prefix separator for Telegram reply

``TELEGRAM_MESSAGE_REPLY_SUFFIX="]"``
    Suffix separator for Telegram reply

``TELEGRAM_MESSAGE_REPLY_LENGTH=15``
    Length of quoted reply message before truncation

``SHOW_TOPIC_MESSAGE=false``
    Send Telegram message when the topic in the IRC channel is changed.

``SHOW_ACTION_MESSAGE=true``
    Relay action messages (e.g. ``/me thinks TeleIRC is cool!``).

``SHOW_JOIN_MESSAGE=false``
    Send Telegram message when someone joins IRC channel.

``JOIN_MESSAGE_ALLOW_LIST=""``
    List of users (separated by a space character) whose IRC leave messages will be sent to Telegram, even if SHOW_JOIN_MESSAGE is false.
    This is ignored if SHOW_JOIN_MESSAGE is set to true.

``SHOW_KICK_MESSAGE=false``
    Send Telegram message when someone is kicked from IRC channel.

``SHOW_NICK_MESSAGE=false``
    Send Telegram message when someone changes their nickname in the IRC channel.

``SHOW_LEAVE_MESSAGE=false``
    Send Telegram message when someone leaves IRC channel either by quitting or parting.

``LEAVE_MESSAGE_ALLOW_LIST=""``
    List of users (separated by a space character) whose IRC leave messages will be sent to Telegram, even if SHOW_LEAVE_MESSAGE is false.
    This is ignored if SHOW_LEAVE_MESSAGE is set to true.

``SHOW_DISCONNECT_MESSAGE=false``
    Sends a message to Telegram when the bot disconnects from the IRC side.

**************
Imgur settings
**************

``IMGUR_CLIENT_ID=7d6b00b87043f58``
    Imgur API client ID value to access Imgur API. Uses a default client ID.
    If you are bridging to a very active Telegram group,
    *please :ref:`register your own application <imgur-setup>`*.

``IMGUR_CLIENT_SECRET=""``
    Imgur API client secret.
    Only needed when :ref:`uploading to an account <imgur-login>`.

``IMGUR_REFRESH_TOKEN=""``
    Imgur API refresh_token for the account where images should be uploaded.
    Only needed when :ref:`uploading to an account <imgur-login>`.

``IMGUR_ALBUM_HASH=""``
    The album hash for the Imgur album uploaded images should belong to.
    :ref:`How to find this <imgur-album>`


internal/config.go
package internal

import (
	"fmt"
	"os"
	"strings"

	"github.com/caarlos0/env/v6"
	"github.com/go-playground/validator"
	"github.com/joho/godotenv"
)

var validate *validator.Validate

const defaultPath = ".env"

// IRCSettings includes settings related to the IRC bot/message relaying
type IRCSettings struct {
	BindAddress         string   `env:"IRC_HOST_IP" envDefault:""`
	Server              string   `env:"IRC_SERVER,required"`
	ServerPass          string   `env:"IRC_SERVER_PASSWORD" envDefault:""`
	Port                int      `env:"IRC_PORT" envDefault:"6667" validate:"min=0,max=65535"`
	TLSAllowSelfSigned  bool     `env:"IRC_CERT_ALLOW_SELFSIGNED" envDefault:"true"`
	TLSAllowCertExpired bool     `env:"IRC_CERT_ALLOW_EXPIRED" envDefault:"true"`
	Channel             string   `env:"IRC_CHANNEL,required" validate:"notempty"`
	ChannelKey          string   `env:"IRC_CHANNEL_KEY" envDefault:""`
	BotIdent            string   `env:"IRC_BOT_IDENT,required" envDefault:"teleirc"`
	BotName             string   `env:"IRC_BOT_REALNAME" envDefault:"Powered by TeleIRC <github.com/rubensadev/teleirc>"`
	BotNick             string   `env:"IRC_BOT_NAME,required" validate:"notempty"`
	SendStickerEmoji    bool     `env:"IRC_SEND_STICKER_EMOJI" envDefault:"true"`
	SendDocument        bool     `env:"IRC_SEND_DOCUMENT" envDefault:"true"`
	Prefix              string   `env:"IRC_PREFIX" envDefault:"<"`
	Suffix              string   `env:"IRC_SUFFIX" envDefault:">"`
	ShowJoinMessage     bool     `env:"IRC_SHOW_JOIN_MESSAGE" envDefault:"true"`
	ShowLeaveMessage    bool     `env:"IRC_SHOW_LEAVE_MESSAGE" envDefault:"true"`
	ShowZWSP            bool     `env:"IRC_SHOW_ZWSP" envDefault:"true"`
	ShowLocationMessage bool     `env:"IRC_SHOW_LOCATION_MESSAGE" envDefault:"false"`
	NickServUser        string   `env:"IRC_NICKSERV_USER" envDefault:""`
	NickServPassword    string   `env:"IRC_NICKSERV_PASS" envDefault:""`
	NickServService     string   `env:"IRC_NICKSERV_SERVICE" envDefault:""`
	EditedPrefix        string   `env:"IRC_EDITED_PREFIX" envDefault:"[EDIT] "`
	MaxMessageLength    int      `env:"IRC_MAX_MESSAGE_LENGTH" envDefault:"400"`
	IRCBlacklist        []string `env:"IRC_BLACKLIST" envDefault:"[]string{}"`
	UseSSL              bool     `env:"IRC_USE_SSL" envDefault:"false"`
	NoForwardPrefix     string   `env:"IRC_NO_FORWARD_PREFIX" envDefault:""`
	QuitMessage         string   `env:"IRC_QUIT_MESSAGE" envDefault:""`
}

// TelegramSettings includes settings related to the Telegram bot/message relaying
type TelegramSettings struct {
	Token                 string   `env:"TELEIRC_TOKEN,required"`
	ChatIDs               []int64  `env:"TELEGRAM_CHAT_ID,required"`
	Prefix                string   `env:"TELEGRAM_MESSAGE_PREFIX" envDefault:"<"`
	Suffix                string   `env:"TELEGRAM_MESSAGE_SUFFIX" envDefault:">"`
	ReplyPrefix           string   `env:"TELEGRAM_MESSAGE_REPLY_PREFIX" envDefault:"["`
	ReplySuffix           string   `env:"TELEGRAM_MESSAGE_REPLY_SUFFIX" envDefault:"]"`
	ReplyLength           int      `env:"TELEGRAM_MESSAGE_REPLY_LENGTH" envDefault:"15"`
	ShowTopicMessage      bool     `env:"SHOW_TOPIC_MESSAGE" envDefault:"false"`
	ShowJoinMessage       bool     `env:"SHOW_JOIN_MESSAGE" envDefault:"false"`
	JoinMessageAllowList  []string `env:"JOIN_MESSAGE_ALLOW_LIST" envDefault:"[]string{}"`
	ShowActionMessage     bool     `env:"SHOW_ACTION_MESSAGE" envDefault:"true"`
	ShowLeaveMessage      bool     `env:"SHOW_LEAVE_MESSAGE" envDefault:"false"`
	LeaveMessageAllowList []string `env:"LEAVE_MESSAGE_ALLOW_LIST" envDefault:"[]string{}"`
	ShowKickMessage       bool     `env:"SHOW_KICK_MESSAGE" envDefault:"false"`
	ShowNickMessage       bool     `env:"SHOW_NICK_MESSAGE" envDefault:"false"`
	ShowDisconnectMessage bool     `env:"SHOW_DISCONNECT_MESSAGE" envDefault:"false"`
	MaxMessagePerMinute   int      `env:"MAX_MESSAGE_PER_MINUTE" envDefault:"20"`
}

// ImgurSettings includes settings related to Imgur uploading for Telegram photos
type ImgurSettings struct {
	ImgurClientID     string `env:"IMGUR_CLIENT_ID" envDefault:"7d6b00b87043f58"`
	ImgurClientSecret string `env:"IMGUR_CLIENT_SECRET" envDefault:""`
	ImgurRefreshToken string `env:"IMGUR_REFRESH_TOKEN" envDefault:""`
	ImgurAccessToken  string ``
	ImgurAlbumHash    string `env:"IMGUR_ALBUM_HASH" envDefault:""`
}

// Settings includes all user-configurable settings for TeleIRC
type Settings struct {
	IRC      IRCSettings
	Telegram TelegramSettings
	Imgur    ImgurSettings
}

func validateEmptyString(fl validator.FieldLevel) bool {
	return fl.Field().String() != ""
}

// ConfigErrors lets us wrap the validator errors in a type we can return
type ConfigErrors []validator.FieldError

func (ce ConfigErrors) Error() string {
	finalStr := ""
	for _, err := range ce {
		switch err.Tag() {
		case "notempty":
			finalStr += fmt.Sprintf("Field %s was an empty string. "+
				"Perhaps you had a # and need to surround the value with \"\"?\n", err.Namespace())
		case "min":
			finalStr += fmt.Sprintf("Field %s failed to validate: %s too small.\n",
				err.Namespace(), err.Param())
		case "max":
			finalStr += fmt.Sprintf("Field %s failed to validate: %s too large.\n",
				err.Namespace(), err.Param())
		default:
			finalStr += fmt.Sprintf("Field %s failed to validate: %s failed %s.\n",
				err.Namespace(), err.Param(), err.Tag())
		}
	}
	return finalStr
}

/*
For environment variables that are meant to be a list separated by whitespace,
we need to do that manually.   This function does just that.
*/
func splitEnvVar(list []string) []string {
	if list == nil {
		return list
	} else if len(list) == 0 {
		return list
	}

	return strings.Split(list[0], " ")
}

/*
LoadConfig loads in the .env file in the provided path (or ".env" by default)
If the user-provided config is valid, return a new Settings struct that contains these settings.
Otherwise, return the error that caused the failure.
*/
func LoadConfig(path string) (*Settings, error) {
	validate = validator.New()
	if err := validate.RegisterValidation("notempty", validateEmptyString); err != nil {
		return nil, err
	}
	// Attempt to load environment variables from path if path was provided
	if path != ".env" && path != "" {
		if _, err := os.Stat(path); !os.IsNotExist(err) {
			if err := godotenv.Load(path); err != nil {
				return nil, err
			}
		}
	} else if _, err := os.Stat(defaultPath); !os.IsNotExist(err) {
		// Attempt to load from defaultPath if defaultPath exists
		if err := godotenv.Load(defaultPath); err != nil {
			return nil, err
		}
	}
	settings := &Settings{}
	if err := env.Parse(settings); err != nil {
		return nil, err
	}
	if err := validate.Struct(settings); err != nil {
		fieldErrs := ConfigErrors{}
		for _, errs := range err.(validator.ValidationErrors) {
			fieldErrs = append(fieldErrs, errs)
		}
		return nil, fieldErrs
	}

	if len(settings.Telegram.ChatIDs) < 1 {
		return nil, fmt.Errorf("TELEGRAM_CHAT_ID must contain at least one chat ID")
	}

	settings.IRC.IRCBlacklist = splitEnvVar(settings.IRC.IRCBlacklist)
	settings.Telegram.JoinMessageAllowList = splitEnvVar(settings.Telegram.JoinMessageAllowList)
	settings.Telegram.LeaveMessageAllowList = splitEnvVar(settings.Telegram.LeaveMessageAllowList)

	return settings, nil
}
